src/honey_badger/epoch_state.rs

```diff
@@ -219,15 +222,35 @@ where
     /// Checks whether the subset has output, and if it does, sends out our decryption shares.
     fn process_subset(&mut self, cs_step: cs::Step<N>) -> Result<Step<C, N>> {
         let mut step = Step::default();
-        let mut cs_outputs = step.extend_with(cs_step, |cs_msg| {
+        let cs_outputs: VecDeque<_> = step.extend_with(cs_step, |cs_msg| {
             MessageContent::Subset(cs_msg).with_epoch(self.epoch)
         });
-        if let Some(cs_output) = cs_outputs.pop_front() {
-            self.subset = SubsetState::Complete(cs_output.keys().cloned().collect());
-            step.extend(self.send_decryption_shares(cs_output)?);
-        }
-        if !cs_outputs.is_empty() {
-            error!("Multiple outputs from a single Subset instance.");
+        for cs_output in cs_outputs {
+            match cs_output {
+                SubsetOutput::Contribution(k, v) => {
+                    step.extend(self.send_decryption_share(k.clone(), &v)?);
+                    self.nodes_found_so_far.insert(k);
+                }
+                SubsetOutput::Done => {
+                    self.subset = SubsetState::Complete(self.nodes_found_so_far.clone());
+
+                    let faulty_shares: Vec<_> = self
+                        .decryption
+                        .keys()
+                        .filter(|id| !self.nodes_found_so_far.contains(id))
+                        .cloned()
+                        .collect();
+                    for id in faulty_shares {
+                        if let Some(DecryptionState::Ongoing(td)) = self.decryption.remove(&id) {
+                            for id in td.sender_ids() {
+                                let fault_kind = FaultKind::UnexpectedDecryptionShare;
+                                step.fault_log.append(id.clone(), fault_kind);
+                            }
+                        }
+                    }
+                    break;
```

Comment:
I think it's better to continue with the `for` loop and remove this `break`. In case there are more outputs in `cs_output`, don't ignore them but rather `panic!` in case the program finds any. This way we would prevent bugs later.